### PR TITLE
Simplify `EditorNode` by removing `gui_base` and `theme_base`

### PR DESCRIPTION
**Pull request by AaronRecord**
Essentially turns EditorNode into `gui_base`. `theme_base` is redundant either way.

From a quick inspection everything seems to working just fine as before ¯\\_(ツ)_/¯
I could be missing something, but I tried changing the theme from the default theme to light theme and everything looks as expected

### CODE REVIEW

File: editor/editor_node.cpp [hunk 1]
```diff
@@ -656,10 +656,6 @@ void EditorNode::_notification(int p_what) {
 			editor_data.clear_edited_scenes();
 		} break;
 
-		case Control::NOTIFICATION_THEME_CHANGED: {
-			scene_tab_add_ph->set_custom_minimum_size(scene_tab_add->get_minimum_size());
-		} break;
-
```
Comment (AaronRecord) [inline editor/editor_node.cpp:662]: I don't think this was meant to be added in the first place... but now that `EditorNode` inherits Control this actually did get called and it caused a `nullptr` crash.
See https://github.com/godotengine/godot/pull/57086/files#r931703285

**Comment by trollodel** [comment]
While this seems an improvement, I still don't understand why editor theme items are fetched in that way.
I think most of them can be fetched in the following 2 ways:
- Use the `Control::get_theme_*`directly instead of using a higher level node. This is what is done for theme items of builtin nodes, if I'm not mistaken.
- Use the editor's Theme resource directly, removing the overhead of the Control machinery when getting theme items.

Just an opinion on that.

**Comment by AaronRecord** [comment]
> While this seems an improvement, I still don't understand why editor theme items are fetched in that way. I think most of them can be fetched in the following 2 ways:
> 
> * Use the `Control::get_theme_*` directly instead of using a higher level node. This is what is done for theme items of builtin nodes, if I'm not mistaken.
> * Use the editor's Theme resource directly, removing the overhead of the Control machinery when getting theme items.
> 
> Just an opinion on that.

Agreed, but that should probably be handled in a future PR

**Comment by AaronRecord** [comment]
I just ran a benchmark and it turns out this speeds up editor startup time by 17%... wasn't expecting that 😀
```py
C:\Users\aaron>hyperfine -iw2 "C:\Users\aaron\Downloads\simplify-editor-node\godot.windows.opt.tools.64.exe --editor --path C:\Users\aaron\Projects\Godot\EmptyProject4.0 --quit" "C:\Users\aaron\Downloads\master\godot.windows.opt.tools.64.exe --editor --path C:\Users\aaron\Projects\Godot\EmptyProject4.0 --quit"
Benchmark 1: C:\Users\aaron\Downloads\simplify-editor-node\godot.windows.opt.tools.64.exe --editor --path C:\Users\aaron\Projects\Godot\EmptyProject4.0 --quit
  Time (mean ± σ):      3.569 s ±  0.010 s    [User: 0.002 s, System: 0.005 s]
  Range (min … max):    3.554 s …  3.589 s    10 runs

  Warning: Ignoring non-zero exit code.

Benchmark 2: C:\Users\aaron\Downloads\master\godot.windows.opt.tools.64.exe --editor --path C:\Users\aaron\Projects\Godot\EmptyProject4.0 --quit
  Time (mean ± σ):      4.186 s ±  0.009 s    [User: 0.002 s, System: 0.005 s]
  Range (min … max):    4.172 s …  4.203 s    10 runs

  Warning: Ignoring non-zero exit code.

Summary
  'C:\Users\aaron\Downloads\simplify-editor-node\godot.windows.opt.tools.64.exe --editor --path C:\Users\aaron\Projects\Godot\EmptyProject4.0 --quit' ran
    1.17 ± 0.00 times faster than 'C:\Users\aaron\Downloads\master\godot.windows.opt.tools.64.exe --editor --path C:\Users\aaron\Projects\Godot\EmptyProject4.0 --quit'
```

**Comment by YuriSizov** [comment]
I think, to be safe, I'd keep the `EditorNode` itself as `Node` and keep _one_ GUI root node with the theme applied to it as its child. And then I'd still re-evaluate every call to `EditorNode::get_singleton()->get_gui_base()` and refactor it to not do that and use theme propagation instead.

**Comment by AaronRecord** [comment]
> I think, to be safe, I'd keep the `EditorNode` itself as `Node` and keep _one_ GUI root node with the theme applied to it as its child. And then I'd still re-evaluate every call to `EditorNode::get_singleton()->get_gui_base()` and refactor it to not do that and use theme propagation instead.

I guess I'm okay with this, but what exactly is the risk with turning EditorNode into a Panel? I feel like it just makes the code simpler by removing a bunch of `get_gui_base()`s.

**Comment by YuriSizov** [comment]
> I guess I'm okay with this, but what exactly is the risk with turning EditorNode into a Panel?

EditorNode is the root of the entire editor tree, and people who use it (or the code that uses it) may not expect it to be a GUI element that works by GUI element's rules. I'd even go further and refactor it to separate it from the GUI part completely, and keep it for the high-level initializations only (like settings and plugins?). I'd expect `EditorInterface` to represent the GUI base, I guess.

That said, I don't have a concrete example to point to and say "ah-ha, it breaks!" 🙃

**Comment by YuriSizov** [comment]
Okay, given that you've said on RC that the observed performance boost is more like a performance regression in the current master, I guess that closes that angle. But I love the idea of simplification here overall. So I'd definitely like to have this PR shaped into something regardless of the performance improvement.

Expanding on the idea of `EditorInterface` representing the GUI part of the editor, I think we can go with that. And that way we still have that reduced number of calls, because `EditorInterface::get_singleton()` is a thing. But `EditorInterface` is a part of `editor_plugin` and is also a Node thus far... Maybe a bigger clean up is in order.

-----

So for now, my suggestion would be to remove `theme_base` and the related method (which incidentally didn't even use `theme_base`). Make `gui_base` the root of all editor GUI and make sure that everything in the editor accesses that. If you can, clean up the elements that use the `EditorNode` singleton instead of accessing the theme. I can see that a lot of places won't be an easy fix, because they call this code from a non-GUI node. I'd be fine with a separate PR to refactor those, but I can see that some can be cleaned up here already.

**Comment by YuriSizov** [comment]
@AaronRecord Do you still want to make a simplified version of this PR? Or should we postpone any changes for now until we're ready for a more substantial refactor of EditorNode? (reduz wants it to happen for 4.1)

**Comment by AaronRecord** [comment]
> @AaronRecord Do you still want to make a simplified version of this PR? Or should we postpone any changes for now until we're ready for a more substantial refactor of EditorNode? (reduz wants it to happen for 4.1)

Yeah that shouldn't take long, I'll see if I can do that this week.

**Comment by fire** [comment]
@YuriSizov were you still in favor of this or was is removing gui_base and theme_base a 4.1 thing now?

**Comment by YuriSizov** [comment]
I think at this point this is going to be done as a part of some bigger reorganization of the codebase. So I'm going to close this PR, but thanks for your contribution nonetheless!

**Pull request closed without merging**